src/plugins/kibana/server/routes/api/index_pattern_endpoints.js

```diff
@@ -0,0 +1,258 @@
+const {templateToPattern, patternToTemplate} = require('../../lib/convert_pattern_and_template_name');
+const removeDeprecatedFieldProps = require('../../lib/remove_deprecated_field_props');
+const indexPatternSchema = require('../../lib/schemas/index_pattern_schema');
+
+export default function (server) {
+
+  let handleESError = function (error) {
+    if (error instanceof esErrors.ConnectionFault ||
+      error instanceof esErrors.ServiceUnavailable ||
+      error instanceof esErrors.NoConnections ||
+      error instanceof esErrors.RequestTimeout) {
+      return Boom.serverTimeout(error);
+    } else if (error instanceof esErrors.Conflict || _.contains(error.message, 'index_template_already_exists')) {
+      return Boom.conflict(error);
+    } else if (error instanceof esErrors[403]) {
+      return Boom.forbidden(error);
+    } else if (error instanceof esErrors.NotFound) {
+      return Boom.notFound(error);
+    } else if (error instanceof esErrors.BadRequest || error instanceof TypeError) {
+      return Boom.badRequest(error);
+    } else {
+      return error;
+    }
+  };
+
+  server.route({
+    path: '/api/kibana/index_patterns',
+    method: 'GET',
+    handler: function (req, reply) {
+      const callWithRequest = server.plugins.elasticsearch.callWithRequest;
+
+      const params = {
+        index: '.kibana',
+        type: 'index-pattern',
+        body: {
+          query: {
+            match_all: {}
+          }
+        }
+      };
+
+      callWithRequest(req, 'search', params)
+        .then(function parseResults(results) {
+          const hits = results.hits.hits;
+          return _.map(hits, (patternHit) => {
+            patternHit._source.fields = JSON.parse(patternHit._source.fields);
+            return patternHit._source;
+          });
+        })
+        .then((patterns) => {
+          return Promise.map(patterns, (pattern) => {
+            return getMappings(pattern.title, req).catch(() => {
+              return {};
+            }).then((mappings) => {
+              return stitchPatternAndMappings(pattern, mappings);
+            });
+          });
+        })
+        .then(removeDeprecatedFieldProps)
+        .then((patterns) => {
+          reply(patterns);
+        }, function (error) {
+          reply(handleESError(error));
+        });
+    }
+  });
+
+  server.route({
+    path: '/api/kibana/index_patterns/{id}',
+    method: 'GET',
+    handler: function (req, reply) {
+      const callWithRequest = server.plugins.elasticsearch.callWithRequest;
+      let pattern = req.params.id;
+
+      const params = {
+        index: '.kibana',
+        type: 'index-pattern',
+        id: req.params.id
+      };
+
+      Promise.join(
+        callWithRequest(req, 'get', params)
+          .then((result) => {
+            result._source.fields = JSON.parse(result._source.fields);
+            return result._source;
+          }),
+        getMappings(pattern, req),
+        stitchPatternAndMappings)
```

Comment:
@spalger I remember you had a comment recommending an alternative to join, but I can't find that comment anymore. Seems like it got deleted somehow. What was your recommended alternative?